### PR TITLE
decomp xEnv.cpp

### PR DESCRIPTION
**Pull request by mattbruv**

### CODE REVIEW

File: src/Core/p2/iEnv.h [hunk 1]
```diff
@@ -1,6 +1,27 @@
 #ifndef IENV_H
 #define IENV_H
 
+extern struct RpWorld;
+extern struct xJSPHeader;
+extern struct RpLight;
+extern struct RwFrame;
```
Comment (DarkRTA) [inline src/Core/p2/iEnv.h]: Opaque struct definitions don't need `extern` unless there's some weird mwcceppc quirk I'm unaware of.

They also don't belong here in the first place but should be in the correct header file.

File: src/Core/p2/iEnv.h [hunk 2]
```diff
@@ -1,6 +1,27 @@
...
+	xJSPHeader* jsp;
+	RpLight* light[2];
+	RwFrame* light_frame[2];
+	int memlvl;
+};
+
+void iEnvFree(iEnv* param_1);
+void iEnvRender(iEnv* param_1);
+void iEnvDefaultLighting(iEnv* param_1);
+//void iEnvLoad(iEnv* env, void* data, int dataType);
```
Comment (DarkRTA) [inline src/Core/p2/iEnv.h]: Is there a reason why this is commented out?

File: src/Core/x/xEnv.cpp [hunk 3]
```diff
@@ -1 +1,38 @@
-#include "xEnv.h"
\ No newline at end of file
+#include "xEnv.h"
+
+extern xEnv* gCurXEnv;
```
Comment (DarkRTA) [inline src/Core/x/xEnv.cpp:4]: I'd leave a todo comment here saying this needs to be moved from xEntMotion.s

File: src/Core/x/xEnv.h [hunk 4]
```diff
@@ -1,7 +1,18 @@
 #ifndef XENV_H
 #define XENV_H
 
-typedef struct xEnv;
+#include "../../dolphin/include/dolphin/types.h"
+#include "../p2/iEnv.h"
+
+extern struct xLightKit;
```
Comment (DarkRTA) [inline src/Core/x/xEnv.h]: Same deal with the first group of opaque structs. This doesn't need extern and should be defined in xLightKit.h.

**Comment by DarkRTA** [comment]
Also you should be using the int types like `s32` from `<dolphin/types.h>` (btw we really need to fix that include path as it doesn't work at the moment)

**Comment by seilc** [comment]
We should not be using the types from the dolphin types header. Those are reserved for the Dolphin SDK only

**Comment by seilc** [comment]
We shouldn't even need to include dolphin headers unless we specifically need a struct or something from them

**Comment by DarkRTA** [comment]
> We should not be using the types from the dolphin types header. Those are reserved for the Dolphin SDK only

This is about the integer types with a known width. If using `dolphin/types.h` is not an option we should clone what we need from it and use our own `types.h`.

`int` can have a different meaning depending on what you compile for whereas something like `s32` is known to be a 32 bit signed int if the typedef is correct.

And the assumption that we are currently targeting GCN is only true at the moment. Later on down the line this codebase is *going* to be ported to other platforms at some point, and that port will have a lot of problems if we aren't using known width integer types.

**Comment by DarkRTA** [comment]
Also, if we aren't supposed to be using the Dolphin SDK `types.h` outside of the Dolphin SDK, then we should *really* fix this include:

https://github.com/pslehisl/bfbbdecomp/blob/b2ebb268b29038e41c4ad2a5ccd12f3507c39d36/src/Game/zParEmitter.cpp#L3

**Comment by DarkRTA** [comment]
Oh yeah i'd also remove xEnv.s if we no longer need it.

**Comment by mattbruv** [comment]
I fully decomped another file just now. What do we need to do to get this merged?

**Comment by DarkRTA** [comment]
Address the issues i pointed out earlier in the thread

**Pull request merged**